Mr 97 : Support for MariaDB

Alexandre-T opened the pull request:
#97 

I continue the excellent job of @PowerKiKi

PowerKiKi committed:
Support for MariaDB 10.6

PowerKiKi committed:
MariaDB does not implements ST_NumInteriorRing, but ST_NumInteriorRings

PowerKiKi committed:
Most tests are passing

PowerKiKi committed:
All `*FromWkb` functions are not compatible with MariaDB

Because we could not find a way to pass a binary parameter correctly.
All 26 combinations of `Doctrine\DBAL\ParameterType::*` and
`Doctrine\DBAL\ParameterType::*`, and with or without `hex2bin()` were
tested. But none are working.

PowerKiKi committed:
Declare spatial types compatibility

PowerKiKi committed:
Code style

PowerKiKi committed:
Tolerate higher coupling between objects

PowerKiKi committed:
Avoid `else` expression

Alexandre-T committed:
Merge branch 'main' into mr-97
Minor updates
Docker file updated

Alexandre-T committed:
Fix SRID function

Alexandre-T committed:
Fix Copyright of MariaDB new classes

Alexandre-T committed:
PHP-CS-FIXER : used classes ordered

Alexandre-T committed:
Fixing typo in ScSRID

Alexandre-T committed:
.gitignore files removed

Alexandre-T committed:
Unused variable

Alexandre-T committed:
helthcheck updated
service added for other platform

Alexandre-T committed:
MariaDB healthcheck removed

Alexandre-T commented:
Errors on `doctrine/orm:4.0.x-dev` come from the walker. Walker have been updated. Issues are NOT linked with this PR.

Errors on `doctrine/orm:2.9`certainly come from the MySQLPlatform class. I really need an object to determine the exact platform. Doctrine team made a mess with this upgrade 😂 . (I'm joking, I certainly didn't understand the process). In any case, I have to create an object and unit test it well with each doctrine version.

Alexandre-T commented:
Hello @PowerKiKi ,

Could you try to run on your local machine this command to force the old version of doctrine:
```
composer --prefer-stable require doctrine/orm:^2.19 --with-all-dependencies
```
Then could you run the test on mariadb? I encounter a lot of error as you can read [in this job](https://github.com/longitude-one/doctrine-spatial/actions/runs/12908573744/job/35994520546?pr=117).

I found how to fix the three last ones, but I don't understand why the first ones failed especially this one:
```
3) LongitudeOne\Spatial\Tests\ORM\Query\AST\Functions\Standard\StGeomFromWkbTest::testSelectWithSrid
Doctrine\DBAL\Exception\DriverException: An exception occurred while executing a query: SQLSTATE[HY000]: General error: 4079 Illegal parameter data type varchar for operation 'st_geometryfromwkb'
```
Is there a link with your explanation with the table you provided?

PowerKiKi commented:
Yes, I also get errors after following your instructions, and yes `Illegal parameter data type varchar` is related to [my table](https://github.com/longitude-one/doctrine-spatial/pull/97#issuecomment-2310456407). I had a look again and cannot come up with a solution either.

My opinion is that since [ORM 3.0 was released one year ago](https://www.doctrine-project.org/2024/02/03/doctrine-orm-3-and-dbal-4-released.html), and since ORM 2.9 will only be supported one more year (until end of 2025 apparently), then we can reasonably drop support for ORM 2.9 too. Or at worst, keep the support declaration, but add some kind of warning in our documentation, keep the production code as-is and edit test code to skip those tests on ORM 2.9. So we can install and use `longitude-one/doctrine-spatial` with ORM 2.9 and MariaDB as long as we never use the `*FromWkb` functions. 

What do you think ?

Alexandre-T commented:
Hello @PowerKiKi ,

Thanks for your reply.

I want to guarantee backward compatibility and not break compatibility with Doctrine ORM 2.9. Compatibility will be break with next major version. This version will use the strategy design pattern which will help us to persist data without loosing SRID. This kind of issues will disappear.

Your solution seems to me an excellent compromise and I will start with that. I think I could buckle down to it on Sunday.

PowerKiKi commented:
Shall I "edit test code to skip those tests on ORM 2.9" (when I have some time :sweat_smile:) ? or is it still on your todo list ?

PowerKiKi commented:
I created #127 to complete this PR.

PowerKiKi committed:
Complete Mr 97 : Support for MariaDB (#127)

* Skip testing of *FromWkb on MariaDB with Doctrine 2.9

* Detect MariaDB first

Because MariaDB inherits from MySQL, so it must be detected first,
otherwise we incorrectly use the branch meant for MySQL when we run
against MariaDB

* Skip failing test on MariaDB and ORM 2.9

* Don't run test against MariaDB if support is declared only for MySQL

Alexandre-T committed:
Add PHP 8.4 support to the testing matrix in GitHub Actions (#128)

Alexandre-T commented:
Great job @PowerKiKi!
